Implement Python wrapper for parity-clib and example

mjkoo opened the pull request:
Follows `jni` example, uses `pyo3` to build an extension module exposing the low-level API and includes a Python module which exports a more pythonic interface.

Ref: https://github.com/paritytech/parity-ethereum/issues/9339 and https://github.com/paritytech/parity-ethereum/issues/8604

Interested in also tackling https://github.com/paritytech/parity-ethereum/issues/9338 (Golang example) if this is well-received.

niklasad1 commented on parity-clib/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ futures = "0.1.6"
 jni = { version = "0.11", optional = true }
 panic_hook = { path = "../util/panic-hook" }
 parity-ethereum = { path = "../", default-features = false }
+pyo3 = { version = "0.6.0-alpha.4", optional = true, features = ["extension-module"] }
```
`pyo3` will require the nightly toolchain and currently we only use stable features

niklasad1 commented on parity-clib/src/python.rs:
```diff
@@ -0,0 +1,219 @@
+
+impl Drop for ConfigHandle {
+	fn drop(&mut self) {
+		if self.inner != ptr::null_mut() {
```
```suggestion
		if !self.inner.is_null() {
```

niklasad1 commented on parity-clib/src/python.rs:
```diff
@@ -0,0 +1,219 @@
+#[pyfunction]
+fn rpc_query(
+	_py: Python<'_>,
+	parity: &ParityHandle,
```
how is it ensured that this function can't end up with `undefined behavior`?

mjkoo replied:
Need some help here, what guarantees should I be ensuring? Is the concern about the behavior of the callback?

Depending on response java code may also need addressing.

niklasad1 replied:
Sorry, if I my comment wasn't clear enough but what I meant was how is it ensured that this function is memory safe (aliasing, dangling pointer or similar)?

My only concern is that this function is not marked as unsafe, can you explain why?

niklasad1 commented on parity-clib/src/python.rs:
```diff
@@ -0,0 +1,219 @@
+	callback: PyObject,
+) -> PyResult<SubscriptionHandle> {
+	if parity.inner == ptr::null_mut() {
+		return Err(ValueError::py_err("Parity execution completed"));
```
hrm, this means that `ParityClient` has been shutdown or it failed during init right?

Can you change the error message something like: "Attempt to subscribe to WebSocket when ParityClient is not running"?

niklasad1 commented on parity-clib/src/python.rs:
```diff
@@ -0,0 +1,219 @@
+	rpc: &str,
+	callback: PyObject,
+) -> PyResult<SubscriptionHandle> {
+	if parity.inner == ptr::null_mut() {
```
```suggestion
	if parity.inner.is_null() {
```

niklasad1 commented on parity-clib/python/parity/__init__.py:
```diff
@@ -0,0 +1,121 @@
+        config = _parity.config_from_cli(options)
+        self.handle = _parity.build(config, logger_mode, logger_file)
+
+    def rpc_query_cb(self, query, cb, timeout_ms=1000):
```
Personally I would rename this to  `rpc_query_async`

niklasad1 commented on parity-clib/python/parity/__init__.py:
```diff
@@ -0,0 +1,121 @@
+        """
+        _parity.rpc_query(self.handle, query, timeout_ms, cb)
+
+    def rpc_query(self, query, timeout_ms=1000):
```
`rpc_query_sync`

niklasad1 commented on parity-clib/python/parity/__init__.py:
```diff
@@ -0,0 +1,121 @@
+        """
+        cb = _CallbackGenerator()
+        self.rpc_query_cb(query, cb, timeout_ms)
+        return next(cb)
```
This will call the iterator of the Callback class?

mjkoo replied:
of the `_CallbackGenerator` class, yes

niklasad1 commented on parity-clib/src/python.rs:
```diff
@@ -0,0 +1,219 @@
+
+use parity_ethereum::RunningClient;
+use pyo3::exceptions::{RuntimeError, ValueError};
+use pyo3::prelude::*;
```
Please use explicit imports

niklasad1 commented on parity-clib/python/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*.so
```
hrm, move this to https://github.com/paritytech/parity-ethereum/blob/master/.gitignore

mjkoo replied:
Already in there, removed from this file. Strange though, pretty sure I added this for a reason :thinking:

mjkoo commented:
@niklasad1 thank you for review, partially addressed comments, had some follow up questions asked above.

Also experienced ICE when compiling with latest nightly, had to revert to nightly-2019-02-07 which is what I've been testing with.

Ack re: pyo3 and stable features, especially since this is triggering bug with current nightly toolchain. Will follow stabilization efforts over there, seems fairly close.

niklasad1 commented:
FYI, the ICE is fixed now https://github.com/rust-lang/rust/issues/58634

debris commented:
I would love to get it in. What is the status of this pr? @niklasad1 do you still have some suggestions regarding this pr?

debris commented:
I would love to get it in. What is the status of this pr? @niklasad1 do you still have some suggestions regarding this pr?

niklasad1 commented:
@debris it is currently blocked on https://github.com/PyO3/pyo3/issues/210 but I agree would be nice to get it in. We shall maybe look how hard it to fix that issue IMHO.

niklasad1 commented:
We have removed `parity-clib`, thus closing this one.
